File: src/lib/stateMachines.ts
/**
 * MIA by VIA — 10 Core Aggregate State Machines Engine
 * Implements strict state transition matrices, authority guards, and validation rules.
 */

// 1. CitizenIdentity States & Triggers
export type CitizenIdentityState =
  | 'DRAFT'
  | 'VERIFICATION_PENDING'
  | 'ACTIVE'
  | 'RESTRICTED'
  | 'SUSPENDED'
  | 'RECOVERY_PENDING'
  | 'CLOSED';

export type CitizenIdentityTrigger =
  | 'SUBMIT_ENROLLMENT'
  | 'APPROVE_VERIFICATION'
  | 'REJECT_VERIFICATION'
  | 'APPLY_RESTRICTION'
  | 'LIFT_RESTRICTION'
  | 'SUSPEND_IDENTITY'
  | 'INITIATE_RECOVERY'
  | 'APPROVE_RECOVERY'
  | 'CLOSE_IDENTITY';

// 2. CitizenNode States & Triggers
export type CitizenNodeState =
  | 'UNBOUND'
  | 'BOUND'
  | 'ACTIVE'
  | 'ROTATION_PENDING'
  | 'RECOVERY_PENDING'
  | 'COMPROMISED'
  | 'REVOKED';

export type CitizenNodeTrigger =
  | 'BIND_DID'
  | 'ACTIVATE_NODE'
  | 'REQUEST_KEY_ROTATION'
  | 'COMPLETE_KEY_ROTATION'
  | 'REPORT_LOST_DEVICE'
  | 'APPROVE_NODE_RECOVERY'
  | 'MARK_COMPROMISED'
  | 'REVOKE_NODE';

// 3. GovernmentCode States & Triggers
export type GovernmentCodeState =
  | 'DRAFT'
  | 'REVIEW'
  | 'AUTHORIZED'
  | 'ACTIVE'
  | 'PAUSED'
  | 'RETIRED';

export type GovernmentCodeTrigger =
  | 'SUBMIT_FOR_REVIEW'
  | 'AUTHORIZE_GCODE'
  | 'ACTIVATE_GCODE'
  | 'PAUSE_GCODE'
  | 'RESUME_GCODE'
  | 'RETIRE_GCODE';

// 4. Credential States & Triggers
export type CredentialState =
  | 'DRAFT'
  | 'PENDING_ISSUANCE'
  | 'ISSUED'
  | 'ACTIVE'
  | 'SUSPENDED'
  | 'REVOKED'
  | 'EXPIRED'
  | 'SUPERSEDED';

export type CredentialTrigger =
  | 'PROPOSE_CREDENTIAL'
  | 'SIGN_AND_DELIVER'
  | 'ACTIVATE_CREDENTIAL'
  | 'SUSPEND_CREDENTIAL'
  | 'UNSUSPEND_CREDENTIAL'
  | 'REVOKE_CREDENTIAL'
  | 'EXPIRE_CREDENTIAL'
  | 'SUPERSEDE_CREDENTIAL';

// 5. ProofRequest States & Triggers
export type ProofRequestState =
  | 'CREATED'
  | 'PRESENTED'
  | 'VERIFIED'
  | 'DECLINED'
  | 'EXPIRED'
  | 'CANCELLED';

export type ProofRequestTrigger =
  | 'PRESENT_PROOF'
  | 'VERIFY_PROOF'
  | 'DECLINE_PROOF'
  | 'EXPIRE_PROOF'
  | 'CANCEL_PROOF';

// 6. AccessGrant States & Triggers
export type AccessGrantState =
  | 'PROPOSED'
  | 'PENDING_CONSENT'
  | 'ACTIVE'
  | 'EXPIRED'
  | 'REVOKED';

export type AccessGrantTrigger =
  | 'REQUEST_CONSENT'
  | 'GRANT_CONSENT'
  | 'EXPIRE_GRANT'
  | 'REVOKE_GRANT';

// 7. ServiceCase States & Triggers
export type ServiceCaseState =
  | 'DRAFT'
  | 'SUBMITTED'
  | 'UNDER_REVIEW'
  | 'ACTION_REQUIRED'
  | 'APPROVED'
  | 'DENIED'
  | 'ISSUED'
  | 'CLOSED'
  | 'APPEALED';

export type ServiceCaseTrigger =
  | 'SUBMIT_CASE'
  | 'START_REVIEW'
  | 'REQUEST_INFO'
  | 'PROVIDE_INFO'
  | 'APPROVE_CASE'
  | 'DENY_CASE'
  | 'ISSUE_CASE_CREDENTIAL'
  | 'CLOSE_CASE'
  | 'APPEAL_CASE';

// 8. CivicValueAccount States & Triggers
export type CivicValueAccountState = 'OPEN' | 'RESTRICTED' | 'SUSPENDED' | 'CLOSED';

// 9. ValueInstruction States & Triggers
export type ValueInstructionState =
  | 'DRAFT'
  | 'VALIDATING'
  | 'PENDING_APPROVAL'
  | 'AUTHORIZED'
  | 'SUBMITTED'
  | 'SETTLING'
  | 'SETTLED'
  | 'FAILED'
  | 'REVERSED'
  | 'DISPUTED';

export type ValueInstructionTrigger =
  | 'VALIDATE_INSTRUCTION'
  | 'REQUIRE_APPROVAL'
  | 'AUTHORIZE_INSTRUCTION'
  | 'SUBMIT_TO_RAIL'
  | 'SETTLE_INSTRUCTION'
  | 'FAIL_INSTRUCTION'
  | 'REVERSE_INSTRUCTION'
  | 'DISPUTE_INSTRUCTION';

// 10. AuditReceipt States & Triggers
export type AuditReceiptState = 'CREATED' | 'SEALED' | 'VERIFIED' | 'EXCEPTION';

// Transition Matrix Specifications
export const CITIZEN_IDENTITY_TRANSITIONS: Record<CitizenIdentityState, Partial<Record<CitizenIdentityTrigger, CitizenIdentityState>>> = {
  DRAFT: { SUBMIT_ENROLLMENT: 'VERIFICATION_PENDING' },
  VERIFICATION_PENDING: { APPROVE_VERIFICATION: 'ACTIVE', REJECT_VERIFICATION: 'DRAFT' },
  ACTIVE: { APPLY_RESTRICTION: 'RESTRICTED', SUSPEND_IDENTITY: 'SUSPENDED', INITIATE_RECOVERY: 'RECOVERY_PENDING', CLOSE_IDENTITY: 'CLOSED' },
  RESTRICTED: { LIFT_RESTRICTION: 'ACTIVE', SUSPEND_IDENTITY: 'SUSPENDED' },
  SUSPENDED: { LIFT_RESTRICTION: 'ACTIVE', CLOSE_IDENTITY: 'CLOSED' },
  RECOVERY_PENDING: { APPROVE_RECOVERY: 'ACTIVE', SUSPEND_IDENTITY: 'SUSPENDED' },
  CLOSED: {}
};

export const CREDENTIAL_TRANSITIONS: Record<CredentialState, Partial<Record<CredentialTrigger, CredentialState>>> = {
  DRAFT: { PROPOSE_CREDENTIAL: 'PENDING_ISSUANCE' },
  PENDING_ISSUANCE: { SIGN_AND_DELIVER: 'ISSUED' },
  ISSUED: { ACTIVATE_CREDENTIAL: 'ACTIVE' },
  ACTIVE: { SUSPEND_CREDENTIAL: 'SUSPENDED', REVOKE_CREDENTIAL: 'REVOKED', EXPIRE_CREDENTIAL: 'EXPIRED', SUPERSEDE_CREDENTIAL: 'SUPERSEDED' },
  SUSPENDED: { UNSUSPEND_CREDENTIAL: 'ACTIVE', REVOKE_CREDENTIAL: 'REVOKED' },
  REVOKED: {},
  EXPIRED: {},
  SUPERSEDED: {}
};

export const GCODE_TRANSITIONS: Record<GovernmentCodeState, Partial<Record<GovernmentCodeTrigger, GovernmentCodeState>>> = {
  DRAFT: { SUBMIT_FOR_REVIEW: 'REVIEW' },
  REVIEW: { AUTHORIZE_GCODE: 'AUTHORIZED' },
  AUTHORIZED: { ACTIVATE_GCODE: 'ACTIVE' },
  ACTIVE: { PAUSE_GCODE: 'PAUSED', RETIRE_GCODE: 'RETIRED' },
  PAUSED: { RESUME_GCODE: 'ACTIVE', RETIRE_GCODE: 'RETIRED' },
  RETIRED: {}
};

export const PROOF_REQUEST_TRANSITIONS: Record<ProofRequestState, Partial<Record<ProofRequestTrigger, ProofRequestState>>> = {
  CREATED: { PRESENT_PROOF: 'PRESENTED', CANCEL_PROOF: 'CANCELLED', EXPIRE_PROOF: 'EXPIRED' },
  PRESENTED: { VERIFY_PROOF: 'VERIFIED', DECLINE_PROOF: 'DECLINED', EXPIRE_PROOF: 'EXPIRED' },
  VERIFIED: {},
  DECLINED: {},
  EXPIRED: {},
  CANCELLED: {}
};

export const VALUE_INSTRUCTION_TRANSITIONS: Record<ValueInstructionState, Partial<Record<ValueInstructionTrigger, ValueInstructionState>>> = {
  DRAFT: { VALIDATE_INSTRUCTION: 'VALIDATING' },
  VALIDATING: { REQUIRE_APPROVAL: 'PENDING_APPROVAL', AUTHORIZE_INSTRUCTION: 'AUTHORIZED', FAIL_INSTRUCTION: 'FAILED' },
  PENDING_APPROVAL: { AUTHORIZE_INSTRUCTION: 'AUTHORIZED', FAIL_INSTRUCTION: 'FAILED' },
  AUTHORIZED: { SUBMIT_TO_RAIL: 'SUBMITTED' },
  SUBMITTED: { SETTLE_INSTRUCTION: 'SETTLED', FAIL_INSTRUCTION: 'FAILED' },
  SETTLING: { SETTLE_INSTRUCTION: 'SETTLED', FAIL_INSTRUCTION: 'FAILED' },
  SETTLED: { REVERSE_INSTRUCTION: 'REVERSED', DISPUTE_INSTRUCTION: 'DISPUTED' },
  FAILED: {},
  REVERSED: {},
  DISPUTED: {}
};

File: rust/src/state_machines.rs
//! 10 Core Aggregate State Machines Specification for MIA by VIA
//! Defines explicit enums, valid state transitions, and authority requirements.

use serde::{Deserialize, Serialize};
use std::fmt;

// ---------------------------------------------------------
// 1. CitizenIdentity
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CitizenIdentityState {
    Draft,
    VerificationPending,
    Active,
    Restricted,
    Suspended,
    RecoveryPending,
    Closed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CitizenIdentityTrigger {
    SubmitEnrollment,
    ApproveVerification,
    RejectVerification,
    ApplyRestriction,
    LiftRestriction,
    SuspendIdentity,
    InitiateRecovery,
    ApproveRecovery,
    CloseIdentity,
}

impl CitizenIdentityState {
    pub fn transition(&self, trigger: CitizenIdentityTrigger) -> Result<Self, &'static str> {
        match (self, trigger) {
            (Self::Draft, CitizenIdentityTrigger::SubmitEnrollment) => Ok(Self::VerificationPending),
            (Self::VerificationPending, CitizenIdentityTrigger::ApproveVerification) => Ok(Self::Active),
            (Self::VerificationPending, CitizenIdentityTrigger::RejectVerification) => Ok(Self::Draft),
            (Self::Active, CitizenIdentityTrigger::ApplyRestriction) => Ok(Self::Restricted),
            (Self::Restricted, CitizenIdentityTrigger::LiftRestriction) => Ok(Self::Active),
            (Self::Active, CitizenIdentityTrigger::SuspendIdentity) => Ok(Self::Suspended),
            (Self::Suspended, CitizenIdentityTrigger::LiftRestriction) => Ok(Self::Active),
            (Self::Active, CitizenIdentityTrigger::InitiateRecovery) => Ok(Self::RecoveryPending),
            (Self::RecoveryPending, CitizenIdentityTrigger::ApproveRecovery) => Ok(Self::Active),
            (Self::Active, CitizenIdentityTrigger::CloseIdentity) => Ok(Self::Closed),
            _ => Err("Invalid state transition for CitizenIdentity"),
        }
    }
}

// ---------------------------------------------------------
// 2. CitizenNode
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CitizenNodeState {
    Unbound,
    Bound,
    Active,
    RotationPending,
    RecoveryPending,
    Compromised,
    Revoked,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CitizenNodeTrigger {
    BindDid,
    ActivateNode,
    RequestKeyRotation,
    CompleteKeyRotation,
    ReportLostDevice,
    ApproveNodeRecovery,
    MarkCompromised,
    RevokeNode,
}

impl CitizenNodeState {
    pub fn transition(&self, trigger: CitizenNodeTrigger) -> Result<Self, &'static str> {
        match (self, trigger) {
            (Self::Unbound, CitizenNodeTrigger::BindDid) => Ok(Self::Bound),
            (Self::Bound, CitizenNodeTrigger::ActivateNode) => Ok(Self::Active),
            (Self::Active, CitizenNodeTrigger::RequestKeyRotation) => Ok(Self::RotationPending),
            (Self::RotationPending, CitizenNodeTrigger::CompleteKeyRotation) => Ok(Self::Active),
            (Self::Active, CitizenNodeTrigger::ReportLostDevice) => Ok(Self::RecoveryPending),
            (Self::RecoveryPending, CitizenNodeTrigger::ApproveNodeRecovery) => Ok(Self::Active),
            (Self::Active, CitizenNodeTrigger::MarkCompromised) => Ok(Self::Compromised),
            (Self::Compromised, CitizenNodeTrigger::RevokeNode) => Ok(Self::Revoked),
            _ => Err("Invalid state transition for CitizenNode"),
        }
    }
}

// ---------------------------------------------------------
// 3. GovernmentCode
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum GovernmentCodeState {
    Draft,
    Review,
    Authorized,
    Active,
    Paused,
    Retired,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum GovernmentCodeTrigger {
    SubmitForReview,
    AuthorizeGCode,
    ActivateGCode,
    PauseGCode,
    ResumeGCode,
    RetireGCode,
}

impl GovernmentCodeState {
    pub fn transition(&self, trigger: GovernmentCodeTrigger) -> Result<Self, &'static str> {
        match (self, trigger) {
            (Self::Draft, GovernmentCodeTrigger::SubmitForReview) => Ok(Self::Review),
            (Self::Review, GovernmentCodeTrigger::AuthorizeGCode) => Ok(Self::Authorized),
            (Self::Authorized, GovernmentCodeTrigger::ActivateGCode) => Ok(Self::Active),
            (Self::Active, GovernmentCodeTrigger::PauseGCode) => Ok(Self::Paused),
            (Self::Paused, GovernmentCodeTrigger::ResumeGCode) => Ok(Self::Active),
            (Self::Active, GovernmentCodeTrigger::RetireGCode) => Ok(Self::Retired),
            _ => Err("Invalid state transition for GovernmentCode"),
        }
    }
}

// ---------------------------------------------------------
// 4. Credential (Verifiable Credential)
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CredentialState {
    Draft,
    PendingIssuance,
    Issued,
    Active,
    Suspended,
    Revoked,
    Expired,
    Superseded,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CredentialTrigger {
    ProposeCredential,
    SignAndDeliver,
    ActivateCredential,
    SuspendCredential,
    UnsuspendCredential,
    RevokeCredential,
    ExpireCredential,
    SupersedeCredential,
}

impl CredentialState {
    pub fn transition(&self, trigger: CredentialTrigger) -> Result<Self, &'static str> {
        match (self, trigger) {
            (Self::Draft, CredentialTrigger::ProposeCredential) => Ok(Self::PendingIssuance),
            (Self::PendingIssuance, CredentialTrigger::SignAndDeliver) => Ok(Self::Issued),
            (Self::Issued, CredentialTrigger::ActivateCredential) => Ok(Self::Active),
            (Self::Active, CredentialTrigger::SuspendCredential) => Ok(Self::Suspended),
            (Self::Suspended, CredentialTrigger::UnsuspendCredential) => Ok(Self::Active),
            (Self::Active, CredentialTrigger::RevokeCredential) => Ok(Self::Revoked),
            (Self::Active, CredentialTrigger::ExpireCredential) => Ok(Self::Expired),
            (Self::Active, CredentialTrigger::SupersedeCredential) => Ok(Self::Superseded),
            _ => Err("Invalid state transition for Credential"),
        }
    }
}

// ---------------------------------------------------------
// 5. ProofRequest
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ProofRequestState {
    Created,
    Presented,
    Verified,
    Declined,
    Expired,
    Cancelled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ProofRequestTrigger {
    PresentProof,
    VerifyProof,
    DeclineProof,
    ExpireProof,
    CancelProof,
}

impl ProofRequestState {
    pub fn transition(&self, trigger: ProofRequestTrigger) -> Result<Self, &'static str> {
        match (self, trigger) {
            (Self::Created, ProofRequestTrigger::PresentProof) => Ok(Self::Presented),
            (Self::Presented, ProofRequestTrigger::VerifyProof) => Ok(Self::Verified),
            (Self::Presented, ProofRequestTrigger::DeclineProof) => Ok(Self::Declined),
            (Self::Created, ProofRequestTrigger::CancelProof) => Ok(Self::Cancelled),
            (Self::Created, ProofRequestTrigger::ExpireProof) => Ok(Self::Expired),
            _ => Err("Invalid state transition for ProofRequest"),
        }
    }
}

// ---------------------------------------------------------
// 6. AccessGrant
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AccessGrantState {
    Proposed,
    PendingConsent,
    Active,
    Expired,
    Revoked,
}

// ---------------------------------------------------------
// 7. ServiceCase
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ServiceCaseState {
    Draft,
    Submitted,
    UnderReview,
    ActionRequired,
    Approved,
    Denied,
    Issued,
    Closed,
    Appealed,
}

// ---------------------------------------------------------
// 8. CivicValueAccount
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CivicValueAccountState {
    Open,
    Restricted,
    Suspended,
    Closed,
}

// ---------------------------------------------------------
// 9. ValueInstruction
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ValueInstructionState {
    Draft,
    Validating,
    PendingApproval,
    Authorized,
    Submitted,
    Settling,
    Settled,
    Failed,
    Reversed,
    Disputed,
}

// ---------------------------------------------------------
// 10. AuditReceipt
// ---------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AuditReceiptState {
    Created,
    Sealed,
    Verified,
    Exception,
}

impl fmt::Display for CitizenIdentityState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

impl fmt::Display for CredentialState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

impl fmt::Display for GovernmentCodeState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}
